### PR TITLE
Plugin architecture

### PR DESCRIPTION
**Pull request by benarmston**
This PR introduces a plugin system and rearchitects some core functionality to use the plugin system.

### Life cycle events

The old `EventProcessor` module has been renamed to `EventDispatcher` and is now responsible for dispathing life cycle events to interested parties.  Currently, only plugins can be an interested party.

Life cycle events include events such as "daemon connected", "job created", "resources deallocated", etc..

Life cycle events and are often cross-concern.  For example a "job created" event requires 1) updating the core domain models; 2) potentially running the scheduler; and 3) saving the scheduler state in some manner.

These are all different concerns responding to the same event.  `EventDispatcher` and the plugin system provides a mechanism for these concerns to be kept separate whilst still responding to the same events.

Responding to an event, could cause new life cycle events happen.  For instance, a "job created" event could result in the scheduling algorithm running and a "resource allocated" event being dispatched.

### Plugin architecture

There is a new `Plugins` class which is configured with the names of the plugins to load.  It loads the plugins and keeps a copy of them in its registry.

Plugins can export an API to be used by the system and other plugins.  They can also export an "events processor" which receives any life cycle events it is interested in.

Some documentation on the plugin api is available in the [dummy plugin](https://github.com/openflighthpc/flight-scheduler-controller/blob/plugins/lib/flight_scheduler/plugins/demo/dummy.rb).

### Current plugins

Four plugins have been written.  Three are core functionality and have been mostly implemented as plugins as a means of dog fooding the plugin system.  The fourth is reasonably core functionality.

* `core/domain_model`: core functionality without which a scheduler just can't work.  Maintains the state of the domain model and communicates with the daemons.
* `core/node_attributes`: additional attributes about nodes.  Currently, includes a node's CPUs, GPUs, and memory.  Serves mostly as a reference for how a generic resource plugin similar to Slurm's GRES might work.
* `core/scheduling`: controls when the scheduling algorithm should run.
* `scheduler_state/filetxt`: persists and loads the state of the scheduler to a text file.

### Connection Processors

The `ConnectionProcessors` are no longer responsible for processing any messages from the daemons.  Instead they receive the message, extract the relevant information and dispatch a event.

Any communications to the daemons remain encapsulated in the `ConnectionProcessors`.

`ConnectionProcessors` has been renamed to `Connections`.  `ProcessorRegistry` has been renamed to `ConnectionRegistry`.

### Current limitations of the plugin system.

It currently does not support configuring which plugins are to be loaded.  Though it should be relatively clear how this could be done.

Due to the lack of configuration above, the scheduler algorithm has not yet been ported to the plugin system.

Other functionality, such as partition script runners may be candidates for porting to a plugin, but these have not yet been considered or completed.

### CODE REVIEW

File: app/models/partition/script_runner.rb [hunk 1]
```diff
@@ -108,14 +108,16 @@ def build_nodes_hash(nodes)
                                       .map(&:job)
                                       .partition { |j| j.partition == partition }
 
+        attrs = FlightScheduler.app.plugins.lookup('core/node_attributes')
```
Comment (WilliamMcCumstie) [inline app/models/partition/script_runner.rb:111]: Assumable this will eventually be defined as `FlightScheduler.app.plugins.node_attributes`?

Then some fancy configuration integration will run behind the scenes to locate `core/node_attributes`?

File: lib/flight_scheduler/event_dispatcher.rb [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    run_plugins(:job_step_completed, job, job_step, node_name)
+  end
+
+  def job_step_failed(job, job_step, node_name)
+    run_plugins(:job_step_failed, job, job_step, node_name)
+  end
+
+  private
+
+  def run_plugins(method, *args, **opts)
```
Comment (WilliamMcCumstie) [inline lib/flight_scheduler/event_dispatcher.rb]: This is probably fine for a first pass, however I would it be better if some async task orchestrated the running of the plugins?

Each plugin can trigger other parts of the event lifecycle. This creates the following two problems:

1. Multiple plugins could trigger the same "event" , or
2. A loop forms between multiple "events".

\* An "event" defined as any method call with the same arguments.

Looking through the code base, I don't think either of the cases _can_ currently occur. However this was already tricky to verify with a fairly straightforward set of plugins.

---

Obviously trying to solve the halting problem is a futile endeavour. It will remain the responsibility of the plugin writer to ensure a loop doesn't occur. However the use of a queue, could help alleviate the problem. Something like:

```
def run_plugins(event, *args)
  @queue ||=[]
  @queue << [event, args]
  Async do
    running = @queue.delete [event, args]
    if running.length > 0
      plugins.select { ... }.each { plugin.send(...) }
    else
      logger.info "Skipping event foo as it has already ran"
    end
  end
end
```

This way an event is only ran once, even if its triggered multiple times. This will be particularly useful when a `run_scheduler` event is added. If need be, it should be possible to force each instance of an event to run (e.g. `force: true` or append a timestamp).

The use of `Async` to schedule the tasks should prevent loops from blocking the main process. The loop will still be ran, but other tasks will be scheduled periodically.

It also provides a secondary `EventDispatcher#queue` for the life-cycle events. This combined with some additional logging should make it a bit easier to detect/debug loop conditions.

---

Note: There maybe a use case for a `loop`, so no attempt to break them has been made. E.g. creating a log of which jobs are running periodically.

**Review by WilliamMcCumstie — COMMENTED** [review]
I haven't reviewed this PR in particular depth due to its length. AFAICS it mostly refactors the existing code base into `plugins`. Assumable all the class names/ methods have been fixed accordingly.

The `plugins` mechanism looks good in principle. As you mentioned, it is missing a configuration mechanism. However this will be fairly straight forward to add.

My main `concern` is the `plugins` make it hard to follow the logic of what runs when. Even with our fairly simple set, I struggled to follow it through. It would be nice if there was a bit more control over the "life cycle event queue". See comment below

File: app/models/partition/script_runner.rb (hunk 1 above)
Comment (benarmston) [reply app/models/partition/script_runner.rb:111]: Perhaps.  I haven't thought that far ahead.

File: lib/flight_scheduler/event_dispatcher.rb (hunk 2 above)
Comment (benarmston) [reply lib/flight_scheduler/event_dispatcher.rb]: The way in which events are dispatched has been reworked.  We now.

1. Have a queue of events that need to be dispatched.
2. Have an `Async` task which loops over the queue and dispatches the events.  It yields between each dispatched event.
3. Have a `job_creation_received` event allowing faster response to the CLI when the job is created.

There are likely still issues with

1. Shutting down whilst the event queue is not empty.
2. Returning to the CLI after the `job_creation_received` event has been added but not dispatched.

Both of those should be simple to rectify and need not take up any more time on this PR.
